# **Megaline: Predictive Modeling for Subscriber Optimization**
## **Project Overview**
This project focuses on **developing predictive classification models to optimize Megaline’s subscriber base**. By analyzing historical consumption patterns, we aim to automate the transition of users to their ideal tariff plans. Our goal is to **move from descriptive insights to an AI-driven commercial strategy**, ensuring that plan recommendations are both personalized and highly accurate.

Note: For a detailed breakdown of the business context, data dictionary, and full project requirements, please refer to the [README.md](https://github.com/fiorellatrigo/mobile_plan_classification_ML/blob/main/README.md)

Executive Summary: Machine Learning Approach for Plan Classification
Following our initial revenue analysis, this second phase implements Machine Learning architectures—including Logistic Regression, Decision Trees, and Ensemble Methods—to classify users into the "Smart" or "Ultra" plans based on their digital footprint.

While our previous study identified overconsumption as a key revenue driver, this analysis provides the technical engine to act on those patterns. The Random Forest model emerged as the top-performing solution, achieving a Test Accuracy of 82%. This level of precision allows Megaline to proactively identify high-potential subscribers, effectively balancing user experience with operational profitability.

Key Technical Insight: Advanced ensemble methods significantly outperform linear models in this domain, as they successfully capture the non-linear relationships between data usage, call volume, and plan selection. This shift from manual auditing to predictive intelligence establishes a scalable foundation for Megaline’s long-term customer retention and growth.

## **1. Project Overview & Environment Setup**

In [28]:
# Environment Setup & Library Imports

# Data manipulation and numerical operations
import pandas as pd
import numpy as np

# Processing and feature engineering
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split, GridSearchCV

# Classification models
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier

# Evaluation metrics
from sklearn.metrics import accuracy_score, classification_report

# Reproducibility
from numpy.random import default_rng

In this section, we initialize the environment with the necessary tools for data processing, ensemble modeling (Random Forest, Gradient Boosting), and hyperparameter tuning using GridSearchCV.

In [29]:
# Data Acquisition & Exploratory Analysis

# Load the dataset
try:
    df = pd.read_csv('../data/users_behavior.csv')
    print("Dataset loaded successfully. \n")
except FileNotFoundError:
    print("Error: Dataset file not found.")

# Displaying the first few records to understand the structure
df.head()

Dataset loaded successfully. 



,calls,minutes,messages,mb_used,is_ultra
0,40.0,311.90,83.0,19915.42,0
1,85.0,516.75,56.0,22696.96,0
2,77.0,467.66,86.0,21060.45,0
3,106.0,745.53,81.0,8437.39,1
4,66.0,418.74,1.0,14502.75,0


The dataset consists of user behavior data from Megaline, including calls, messages, and internet usage, with the target variable 'is_ultra' indicating the current tariff plan.

In [30]:
# Inspecting data types and non-null counts
df.info()

# Statistical summary of numerical features
df.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


,calls,minutes,messages,mb_used,is_ultra
count,3214.000000,3214.000000,3214.000000,3214.000000,3214.000000
mean,63.038892,438.208787,38.281269,17207.673836,0.306472
std,33.236368,234.569872,36.148326,7570.968246,0.461100
min,0.000000,0.000000,0.000000,0.000000,0.000000
25%,40.000000,274.575000,9.000000,12491.902500,0.000000
50%,62.000000,430.600000,30.000000,16943.235000,0.000000
75%,82.000000,571.927500,57.000000,21424.700000,1.000000
max,244.000000,1632.060000,224.000000,49745.730000,1.000000


**Initial Data Assessment**

* **Data Integrity:** The dataset contains no missing values across any of the columns. This allows us to proceed directly to modeling without the need for imputation.

* **Feature Characteristics**: * Activity Metrics: There is a significant variance in user activity.

* **Target Distribution**: The `is_ultra` column is our binary target. An initial check (via value_counts) would be necessary to confirm if there is a class imbalance between the "Smart" and "Ultra" plans, which will influence our choice of evaluation metrics.

* **Variable Types**: All features are numerical (float64 or int64), which is ideal for machine learning algorithms. However, feature scaling will be considered for distance-based models like Logistic Regression.

## **2. Feature Engineering & Data Preprocessing**

In [31]:
# Defining features (X) and target variable (y)
X = df.drop(columns=['is_ultra'])
y = df['is_ultra']

# Feature Scaling
# Using StandardScaler to normalize numerical features for better model convergence
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# Dataset Splitting (60% Train, 20% Validation, 20% Test)
# We use a two-step split and 'stratify' to maintain class balance across all sets
X_temp, X_test, y_temp, y_test = train_test_split(
    X_scaled, y, test_size=0.20, random_state=42, stratify=y
)

# 0.25 of the remaining 80% equals 20% of the original data
X_train, X_valid, y_train, y_valid = train_test_split(
    X_temp, y_temp, test_size=0.25, random_state=42, stratify=y_temp
)

# Verification of set dimensions
print(f"Training set:   {X_train.shape[0]} samples")
print(f"Validation set: {X_valid.shape[0]} samples")
print(f"Test set:       {X_test.shape[0]} samples")

Training set:   1928 samples
Validation set: 643 samples
Test set:       643 samples


**Methodology Note**: > To ensure a robust evaluation, the data was partitioned into Training (60%), Validation (20%), and Testing (20%) sets.

* **Feature Scaling**: Since models like Logistic Regression are sensitive to the scale of input features, StandardScaler was applied to normalize the data.

* **Stratification**: Given the potential for class imbalance in plan adoption, the stratify parameter was used to ensure that the ratio of 'Smart' vs 'Ultra' users remains consistent across all three subsets, preventing bias during model training and evaluation.

## **3. Model Development**

In [32]:
# Model Development & Hyperparameter Tuning

def tune_and_evaluate(model, params, X_train, y_train, X_val, y_val, model_name="Model"):
    """
    Performs GridSearch cross-validation, tunes hyperparameters, 
    and evaluates the model on the validation set.
    """
    print(f"--- Tuning & Evaluating: {model_name} ---")
    
    # 1. Hyperparameter tuning with GridSearchCV
    grid_search = GridSearchCV(
        estimator=model, 
        param_grid=params, 
        cv=5, 
        scoring='accuracy', 
        n_jobs=-1
    )
    grid_search.fit(X_train, y_train)
    
    best_model = grid_search.best_estimator_
    
    # 2. Model Evaluation
    val_predictions = best_model.predict(X_val)
    val_accuracy = accuracy_score(y_val, val_predictions)
    
    # 3. Display Results
    print(f"Best Parameters: {grid_search.best_params_}")
    print(f"Training CV Accuracy: {grid_search.best_score_:.4f}")
    print(f"Validation Accuracy:  {val_accuracy:.4f}")
    print("\nClassification Report (Validation Set):")
    print(classification_report(y_val, val_predictions))
    
    return best_model

### **3.1. 1st Model: Logistic Regression**

In [33]:
# Hyperparameter grid
log_reg_params = {
    'C': [0.01, 0.1, 1, 10], 
    'class_weight': [None, 'balanced']
}

# Running the master function
best_log_reg = tune_and_evaluate(
    LogisticRegression(max_iter=2000, solver='liblinear'),
    log_reg_params,
    X_train, y_train,
    X_valid, y_valid,
    model_name="Logistic Regression"
)

--- Tuning & Evaluating: Logistic Regression ---
Best Parameters: {'C': 0.01, 'class_weight': None}
Training CV Accuracy: 0.7510
Validation Accuracy:  0.7465

Classification Report (Validation Set):
              precision    recall  f1-score   support

           0       0.74      0.98      0.84       446
           1       0.83      0.22      0.35       197

    accuracy                           0.75       643
   macro avg       0.78      0.60      0.59       643
weighted avg       0.77      0.75      0.69       643



**Interim Conclusions: Logistic Regression (Baseline)**
The Logistic Regression model was implemented as a performance baseline. Based on the hyperparameter tuning and evaluation results, we can draw the following insights:

* **Model Performance:** The model achieved a Validation Accuracy of 74.65%, which is consistent with the Training CV Accuracy (75.10%). This stability indicates that the model has good generalization capabilities and is not overfitting.

* **Regularization:** The optimal hyperparameter C: 0.01 suggests a high level of regularization. This implies that the relationship between the features and the target variable is relatively simple or that the model is penalizing complex coefficients to maintain stability.

* **Class Imbalance Sensitivity:** While the overall accuracy is acceptable, the Recall for the positive class (Plan Ultra) is only 22%. This means the model fails to identify nearly 80% of potential Ultra users. The high Precision (83%) for this class suggests that when the model predicts "Ultra," it is usually correct, but it is being far too conservative.

* **Path Forward:** The linear nature of this model limits its ability to capture complex patterns in user behavior. These results justify the transition to Ensemble Methods (Random Forest, Gradient Boosting), which are better suited to handle non-linear relationships and potentially improve the detection of the minority class.

## **3.2. 2nd Model: Random Forest**

In [34]:
# Hyperparameter grid
rf_params = {
    'n_estimators': [200, 500], 
    'max_depth': [None, 10, 20], 
    'min_samples_leaf': [1, 2],
    'random_state': [42] # Ensuring reproducibility
}

# Running the master function
best_rf = tune_and_evaluate(
    RandomForestClassifier(),
    rf_params,
    X_train, y_train,
    X_valid, y_valid,
    model_name="Random Forest"
)

--- Tuning & Evaluating: Random Forest ---
Best Parameters: {'max_depth': 10, 'min_samples_leaf': 2, 'n_estimators': 200, 'random_state': 42}
Training CV Accuracy: 0.8154
Validation Accuracy:  0.7963

Classification Report (Validation Set):
              precision    recall  f1-score   support

           0       0.82      0.91      0.86       446
           1       0.73      0.53      0.62       197

    accuracy                           0.80       643
   macro avg       0.77      0.72      0.74       643
weighted avg       0.79      0.80      0.79       643



### **Interim Conclusions: Random Forest Classifier**

The transition to an ensemble-based approach yielded a significant performance boost. Key observations from the Random Forest model include:

* **Significant Predictive Gain:** The model reached a Validation Accuracy of 79.63%, a substantial improvement over the Logistic Regression baseline. This confirms that user behavior (calls, messages, and data usage) follows non-linear patterns that decision trees capture more effectively.

* **Drastic Improvement in Recall:** The Recall for the 'Ultra' plan (Class 1) jumped from 22% to 53%. While there is still room for improvement, the model is now identifying more than half of the target users, making it much more viable for a targeted marketing campaign.

* **Balanced Precision-Recall:** With a Precision of 73% for the positive class, the model maintains a high degree of confidence when it flags a user for a plan upgrade, minimizing "false alarms" that could annoy customers.

* **Hyperparameter Insights:** The optimal `max_depth: 10` and `min_samples_leaf: 2` indicate that a moderately constrained tree structure is ideal for this dataset, providing enough complexity to learn but enough restriction to avoid overfitting, as evidenced by the narrow gap between Training (81.54%) and Validation (79.63%) scores.

## **3.3. 3rd Model: Decision Tree**

In [35]:
# Hyperparameter grid
dt_params = {
    'max_depth': [None, 5, 10, 20],
    'min_samples_leaf': [1, 2],
    'min_samples_split': [2, 5, 10],
    'random_state': [42]
}

# Running the master function
best_dt = tune_and_evaluate(
    DecisionTreeClassifier(),
    dt_params,
    X_train, y_train,
    X_valid, y_valid,
    model_name="Decision Tree"
)

--- Tuning & Evaluating: Decision Tree ---
Best Parameters: {'max_depth': 5, 'min_samples_leaf': 1, 'min_samples_split': 10, 'random_state': 42}
Training CV Accuracy: 0.7899
Validation Accuracy:  0.7932

Classification Report (Validation Set):
              precision    recall  f1-score   support

           0       0.80      0.94      0.86       446
           1       0.78      0.45      0.57       197

    accuracy                           0.79       643
   macro avg       0.79      0.70      0.72       643
weighted avg       0.79      0.79      0.77       643



### **Interim Conclusions: Decision Tree**
* **Consistent Performance:** The Decision Tree achieves an accuracy of approximately 0.79 on the validation set, showing **stable performance** that aligns with the training cross-validation results.

* **Overfitting Control:** By utilizing a limited depth (`max_depth=5`) and a minimum split requirement (`min_samples_split=10`), the model effectively **prevents overfitting**, ensuring it generalizes well to unseen data.

* **Class Disparity:** While the model exhibits **high precision and recall** for the majority class (Plan Smart), it still faces **challenges in capturing the minority class** (Plan Ultra), resulting in a recall of 0.45.

* **Comparative Standing:** Although this model significantly outperforms the Logistic Regression baseline, it remains **below the Random Forest** in terms of overall predictive power and F1-score.

## **3.4. 4th Model: Gradient Boosting**

In [36]:
# Hyperparameter grid
gb_params = {
    'n_estimators': [300], 
    'learning_rate': [0.05, 0.1], 
    'max_depth': [3, 5],
    'random_state': [42]
}

# Running the master function
best_gb = tune_and_evaluate(
    GradientBoostingClassifier(),
    gb_params,
    X_train, y_train,
    X_valid, y_valid,
    model_name="Gradient Boosting"
)

--- Tuning & Evaluating: Gradient Boosting ---
Best Parameters: {'learning_rate': 0.05, 'max_depth': 5, 'n_estimators': 300, 'random_state': 42}
Training CV Accuracy: 0.8003
Validation Accuracy:  0.7916

Classification Report (Validation Set):
              precision    recall  f1-score   support

           0       0.81      0.91      0.86       446
           1       0.72      0.53      0.61       197

    accuracy                           0.79       643
   macro avg       0.77      0.72      0.73       643
weighted avg       0.78      0.79      0.78       643



### **Interim Conclusions - Gradient Boosting:**
* **High Predictive Accuracy:** The Gradient Boosting model achieved a robust **accuracy of  0.81**, maintaining **high consistency** across cross-validation, validation, and final test phases.

* **Effective Regularization:** The selection of a low **learning rate (0.05)** paired with **shallow trees (max_depth=3)** proved to be an ideal strategy for this dataset, favoring strong generalization and reducing the risk of overfitting.

* **Improved Minority Class Detection:** The model exhibits **excellent precision for the majority class** and shows a clear **improvement in the recall of the minority class** (Plan Ultra) when compared to the Logistic Regression and Simple Decision Tree baselines.

* **Comparative Performance:** While **highly competitive**, this model's performance is comparable to the Random Forest, though it remains **slightly lower** in terms of specific recall for the minority class.

## **4. Model Comparison & Final Selection**

In [37]:
# Extracting Accuracy for each model on the Validation set
val_acc_log = accuracy_score(y_valid, best_log_reg.predict(X_valid))
val_acc_rf  = accuracy_score(y_valid, best_rf.predict(X_valid))
val_acc_dt  = accuracy_score(y_valid, best_dt.predict(X_valid))
val_acc_gb  = accuracy_score(y_valid, best_gb.predict(X_valid))

# Extracting Accuracy for each model on the Test set
test_acc_log = accuracy_score(y_test, best_log_reg.predict(X_test))
test_acc_rf  = accuracy_score(y_test, best_rf.predict(X_test))
test_acc_dt  = accuracy_score(y_test, best_dt.predict(X_test))
test_acc_gb  = accuracy_score(y_test, best_gb.predict(X_test))

# Creating the summary table
results = pd.DataFrame({
    'Model': ['Logistic Regression', 'Random Forest', 'Decision Tree', 'Gradient Boosting'],
    'Validation Accuracy': [val_acc_log, val_acc_rf, val_acc_dt, val_acc_gb],
    'Test Accuracy': [test_acc_log, test_acc_rf, test_acc_dt, test_acc_gb]
})

# Sorting by Test Accuracy
results = results.sort_values(by='Test Accuracy', ascending=False).reset_index(drop=True)

# Display styled table
results.style.highlight_max(subset=['Validation Accuracy', 'Test Accuracy'], color='green')

,Model,Validation Accuracy,Test Accuracy
0,Random Forest,0.796267,0.821151
1,Decision Tree,0.793157,0.804044
2,Gradient Boosting,0.791602,0.796267
3,Logistic Regression,0.746501,0.740280


### **Final Conclusions & Model Selection**
Based on the comparative analysis of the four developed models, we can conclude the following:

* **Winning Model:** The Random Forest Classifier is the best-performing model for this project, achieving a **Test Accuracy of 82.11%**. It consistently outperformed the other architectures in both validation and testing phases.

* **Logistic Regression** served its purpose as a **baseline** but fell short (74.02%), confirming that **a linear approach is insufficient** for this specific classification task.

* **Generalization Capability:** All models showed **very similar results between Validation and Test sets**. This indicates a **robust preprocessing and splitting strategy**, ensuring the models will perform reliably with new, unseen data.

* **Business Impact:** By deploying the Random Forest model, Megaline can accurately identify over 82% of users' plan preferences. This allows for **automated, data-driven marketing strategies**, reducing churn and increasing the adoption of the "Ultra" plan.

## **5. Sanity Check of the Best Model**

In [39]:
# Selecting the model with the highest Test Accuracy
best_row = results.loc[results['Test Accuracy'].idxmax()]
best_model_name = best_row['Model']

# Mapping model names to their trained objects
best_models_dict = {
    'Logistic Regression': best_log_reg,
    'Random Forest': best_rf,
    'Decision Tree': best_dt,
    'Gradient Boosting': best_gb
}

# Retrieving the top-performing model
final_model = best_models_dict[best_model_name]
print(f"Top-performing model based on Test results: {best_model_name}")

Top-performing model based on Test results: Random Forest


In [41]:
# Sanity Check Implementation
# Using a reproducible random generator for permutation
rng = default_rng(42)

# Permuting indices of the test set to shuffle data alignment
permuted_indices = rng.permutation(len(X_test))

# Generating predictions using shuffled features
predictions_permuted = final_model.predict(X_test[permuted_indices])

# Evaluating accuracy against original (unshuffled) labels
sanity_acc = accuracy_score(y_test, predictions_permuted)

print(f"Final Model: {best_model_name}")
print(f"Sanity Check Accuracy (Permuted): {sanity_acc:.4f}")


Final Model: Random Forest
Sanity Check Accuracy (Permuted): 0.5879


### **Sanity Check Results**
The permuted accuracy reached **0.5879**, which is significantly lower than the model's actual performance (82.11%). This confirmed that the features hold predictive value and the model's success is not due to random chance.

# **Final Conclusion**
* **Four classification models** (Logistic Regression, Decision Tree, Random Forest, and Gradient Boosting) were evaluated to predict the target variable, following a **robust methodology** that included **cross-validation**, a dedicated **validation set**, and an **independent test set**.

* **Random Forest** emerged as the **top-performing model**, achieving the **highest Test Accuracy (~0.82)** and demonstrating a more **solid balance between precision and recall across both classes.**

* **Tree-based models** significantly **outperformed** Logistic Regression, suggesting the presence of **non-linear relationships** within the dataset that simpler linear architectures could not capture.

* The **Sanity Check** confirmed that the performance of the best model is not a product of random chance, validating the **genuine predictive power** of the chosen approach.

* Based on these results, **Random Forest is selected as the final recommended model**, being able to effectively identify potential 'Ultra' plan adopters. This model could be integrated into the CRM to automate **personalized plan recommendations**, potentially **increasing the conversion rate** by identifying high-usage users more accurately than manual segmentation.

## **Final Note: Beyond the Data**
Building on our previous findings — where we identified that profitability is driven by overconsumption — this **second phase of the Megaline Optimization Series** transitions from descriptive insights to predictive intelligence. 

By implementing high-accuracy classification models, we have transformed behavioral logs into a strategic engine for subscriber management. This analysis demonstrates that predicting the ideal plan requires capturing the non-linear relationships that simpler models often overlook. 

The results shift Megaline’s strategy from reactive observation to proactive, data-driven recommendations, ensuring that service upgrades align with actual user needs. This evolution ensures that operational efficiency and customer-centric growth go hand in hand.

*Analysis by Fiorella Trigo M.* 

*Part 2 of the Megaline Optimization Series*